code/__defines/ZAS.dm

```diff
@@ -7,3 +7,44 @@
 #define BLOCKED 3
 
 #define ZONE_MIN_SIZE 14 //zones with less than this many turfs will always merge, even if the connection is not direct
+
+#define CANPASS_ALWAYS 1
+#define CANPASS_DENSITY 2
+#define CANPASS_PROC 3
+#define CANPASS_NEVER 4
+
+#define ATMOS_CANPASS_TURF(ret,A,B) \
+	if (A.blocks_air & AIR_BLOCKED || B.blocks_air & AIR_BLOCKED) { \
+		ret = BLOCKED; \
+	} \
+	else { \
+		if (B.z != A.z) { \
+			if (!istype(A, /turf/simulated/open)) { \
+				ret = BLOCKED; \
+			} \
+		} \
+		else if (A.blocks_air & ZONE_BLOCKED || B.blocks_air & ZONE_BLOCKED) { \
+			ret = (A.z == B.z) ? ZONE_BLOCKED : AIR_BLOCKED; \
+		} \
+		else if (A.contents.len) { \
+			for (var/thing in A) { \
+				var/atom/movable/AM = thing; \
+				switch (AM.atmos_canpass) \
+					if (CANPASS_ALWAYS) \
+						continue; \
+					if (CANPASS_DENSITY) \
+						if (AM.density) { \
+							ret |= AIR_BLOCKED; \
+						} \
+					if (CANPASS_PROC) { \
+						ret |= AM.c_airblock(B); \
+					} \
+					if (CANPASS_NEVER) { \
+						ret = BLOCKED; \
+					} \
+				if (ret == BLOCKED) { \
```

Comment:
Here's hoping `c_airblock()` never returns `BLOCKED`. Otherwise, this check will not work, as the field might already be polluted.